src/main.rs

```diff
@@ -1,3 +1,200 @@
+    jobs: usize
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Person<'a>(Cow<'a, str>);
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Token<'a>(Cow<'a, str>);
+
+impl<'a> From<&'a str> for Token<'a> {
+    fn from(value: &'a str) -> Self {
+        Token(value.into())
+    }
+}
+impl<'a> From<String> for Token<'a> {
+    fn from(value: String) -> Self {
+        Token(value.into())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Chat<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    chat_type: ChatType,
+    id: u128,
+    messages: Vec<Message<'a>>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+enum ChatType {
+    PublicChannel,
+    PrivateChannel,
+    PublicSupergroup,
+    PrivateSupergroup,
+    PersonalChat,
+    ChatForbidden,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Message<'a> {
+    id: u64,
+    #[serde(rename = "type")]
+    message_type: MessageType,
+    date: NaiveDateTime,
+    from: Option<Person<'a>>,
+    text_entities: Vec<TextEntity>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+enum MessageType {
+    Service,
+    Message,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct TextEntity {
+    #[serde(rename = "type")]
+    _type: TextEntityType,
+    text: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "snake_case")]
+enum TextEntityType {
+    Pre,
+    Bold,
+    Link,
+    Code,
+    Email,
+    Plain,
+    Phone,
+    Italic,
+    Cashtag,
+    Spoiler,
+    Mention,
+    Hashtag,
+    TextLink,
+    Underline,
+    BotCommand,
+    CustomEmoji,
+    MentionName,
+    Strikethrough,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ChatStatistics<'a> {
+    num_tokens: usize,
+    #[serde(borrow)]
+    members_tokens_map: HashMap<Person<'a>, HashMap<Token<'a>, usize>>,
+    #[serde(borrow)]
+    tokens_map: HashMap<Token<'a>, usize>,
+}
+impl<'a> ChatStatistics<'a> {
+    fn gather(chat: &'a Chat, jobs: usize) -> Self {
+        let num_messages = chat.messages.len();
+        let messages_per_thread = num_messages / jobs;
+        let (sender, receiver) = std::sync::mpsc::sync_channel(jobs);
+
+        let mut tokens_map = HashMap::new();
+        let mut members_tokens_map = HashMap::new();
+
+        let iter = chat.messages.iter();
+        std::thread::scope(|s| {
+            for i in 0..jobs {
+                let iter = iter.clone();
+                let sender = sender.clone();
+                s.spawn(move || {
+                    let mut chunk_tokens_map = HashMap::new();
+                    let mut chunk_members_tokens_map: HashMap<Person<'_>, HashMap<Token<'_>, usize>> = HashMap::new();
+
+                    for message in iter.skip(i * messages_per_thread).take(messages_per_thread).filter(|message| message.message_type != MessageType::Service) {
```

Comment:
I suppose that you can do filtering before you split the job. That way you'll always get equal amount of work per thread

Reply:
You mean computing number of messages per thread with filtered messages?